File: frontend/src/api/notifications.ts
import client from "./client";
import type { Notification } from "../types";

export async function listNotifications(unreadOnly = false): Promise<Notification[]> {
  const { data } = await client.get("/notifications", { params: { unread_only: unreadOnly } });
  return data;
}

export async function markRead(notificationId: number): Promise<Notification> {
  const { data } = await client.post(`/notifications/${notificationId}/read`);
  return data;
}

export async function markAllRead(): Promise<{ updated: number }> {
  const { data } = await client.post("/notifications/read-all");
  return data;
}

File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.deps import get_current_user
from app.models.user import User
from app.schemas.notification import NotificationResponse
from app.services import notification_service

router = APIRouter(prefix="/api/v1/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationResponse])
def list_notifications(
    unread_only: bool = Query(False),
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return notification_service.list_notifications(db, user, unread_only)


@router.post("/{notification_id}/read", response_model=NotificationResponse)
def mark_read(notification_id: int, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        return notification_service.mark_read(db, user, notification_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/read-all")
def mark_all_read(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return {"updated": notification_service.mark_all_read(db, user)}
